addons/udes_security/controllers/main.py

```diff
@@ -5,12 +5,100 @@
 
 from odoo import http, SUPERUSER_ID, _
 from odoo.http import request
+from odoo.exceptions import AccessError, ValidationError
 
-from odoo.addons.web.controllers.main import Binary, serialize_exception
+from odoo.addons.web.controllers.main import Binary, serialize_exception, Action
 
 _logger = logging.getLogger(__name__)
 
 
+class SecureAction(Action):
+    @http.route()
+    def load(self, action_id, additional_context=None):
+        """
+        Extend load to check menu level access rights before returning the super call,
+        to prevent users without sufficient permission accessing pages via direct URL
+        """
+        ResUsers = request.env["res.users"]
+
+        menu = self._get_menu_from_action_id(action_id)
+        action = menu.action
+        user = ResUsers.sudo().browse(request.session.uid)
+        # Check if user has access to the menu, trigger AccessError and log a warning if not
+        if not self._user_has_access_right(action, menu, user):
+            _logger.warning(
+                f"User id: {user.id}, name: '{user.name}' had access blocked to action id: {action_id}, name: '{action.name if action else 'N/A'}'"
+            )
+            raise AccessError(_("You do not have permission to access this action"))
+        return super().load(action_id, additional_context=additional_context)
+
+    def _get_menu_from_action_id(self, action_id):
+        """
+        Get the ir.ui.menu(x,) recordset from a given action_id
+        `IrUiMenu.sudo().search([("action.id", "=", action_id)])`
+        does not work as `action` is a Reference field. The other option would be to
+        fetch all ir.ui.menu records and then `.filtered()`, but this is a little slow,
+        so instead do a direct query to get the relevant id.
+
+        All models on the `action` Reference field on `ir.ui.menu` have the same
+        `_sequence` parameter, so there is no risk of colliding ids from separate models.
+
+        They are stored in the db like "ir.actions.act_window,90" or "ir.actions.server,242"
+        due to being a Reference field, so perform a fuzzy search on %,<id>% to get it
+
+        Be extra safe and ensure `action_id` is an int, as if the user enters text in the URL
+        for action= param, it will come through to here as a string
+
+        :param: action_id: int() for an action (ir.actions.server, ir.actions.act_window, etc..)
+
+        :return: ir.ui.menu(x,) recordset, will be empty if a non existent action_id is passed in
+        """
+        IrUiMenu = request.env["ir.ui.menu"]
+
+        if not isinstance(action_id, int):
+            raise ValidationError("Invalid action ID specified.")
+        # Escape the leading and trailing static % symbols with %,
+        # as we don't want to string format them in the execute() call
+        query = "SELECT id FROM ir_ui_menu WHERE action LIKE '%%,%s%%'"
+        request.env.cr.execute(query, (action_id,))
```

Comment:
Do you think this would be easier to parse visually?
```python
query = "SELECT id FROM ir_ui_menu WHERE action LIKE '%%,%s%%'" 
request.env.cr.execute(query, (f'%{action_id}%',))
```

Reply:
I'm happy to f-string it (originally I had a similar thing), but I had the impression that string formatting outside of the .execute() call was generally frowned upon for security

Reply:
it can come from the user, as it's whatever is against `action=` in the URL bar, but that's why I added the type check above (because odoo core doesn't even type check it)

Reply:
Actually on balance I prefer your original approach, as that will blow up if action_id isn't an int.  Best err on the side of safety with this sort of thing.